discovery/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,266 @@
+// re-registering it.
+func (conn *ZooKeeper) SendHeartbeat(service *discovery.ServiceDefinition) {
+	if err := conn.registerService(service); err != nil {
+		log.Warnf("Error registering service %s: %s", service.Name, err)
+	}
+}
+
+func (conn *ZooKeeper) parentPath(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, service.Name)
+}
+
+func (conn *ZooKeeper) getNodeKey(service *discovery.ServiceDefinition) string {
+	return fmt.Sprintf("%s/%s", conn.parentPath(service), service.ID)
+}
+
+func (conn *ZooKeeper) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", conn.Prefix, appName)
+}
+
+var zookeeperUpstreams = make(map[string][]ServiceNode)
+
+// CheckForUpstreamChanges checks another zookeeper node for changes
+func (conn *ZooKeeper) CheckForUpstreamChanges(backendName, backendTag string) bool {
+	// TODO: is there a way to filter by tag in zookeeper?
+	services, err := conn.getServices(backendName)
+	if err != nil {
+		log.Errorf("Failed to query %v: %s", backendName, err)
+		return false
+	}
+	didChange := zookeeperCompareForChange(zookeeperUpstreams[backendName], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		zookeeperUpstreams[backendName] = services
+	}
+	return didChange
+}
+
+func (conn *ZooKeeper) getServices(appName string) ([]ServiceNode, error) {
+	var services []ServiceNode
+
+	key := conn.getAppKey(appName)
+	children, _, error := conn.Client.Children(key)
+	if error != nil {
+		return services, error
+	}
+	for i := range children {
+		path := fmt.Sprintf("%s/%s", key, children[i])
+		data, _, error := conn.Client.Get(path)
+		if error != nil {
+			return services, error
+		}
+		srv, error := decodeZooKeeperNodeValue(data)
+		if error != nil {
+			return services, error
+		}
+		services = append(services, srv)
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func zookeeperCompareForChange(existing, new []ServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByZooKeeperServiceID(existing))
+	sort.Sort(ByZooKeeperServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (conn ZooKeeper) createParentPath(path string) error {
+	pathElements := strings.Split(path, "/")[1:]
+	sep := "/"
+	newPath := ""
+	for i := range pathElements {
+		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
+		if _, err := conn.Client.Create(
+			newPath,
+			nil,
+			0,
+			zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
+			return err
+		}
+	}
+	return nil
+}
+
+func (conn ZooKeeper) registerService(service *discovery.ServiceDefinition) error {
```

Comment:
Not sure I understand your question (sorry for being dense), but no, we don't create any additional state, I think. 

> Il giorno 12 set 2016, alle ore 16:09, Tim Gross <notifications@github.com> ha scritto:
> 
> In discovery/zookeeper/zookeeper.go:
> 
> > +	sep := "/"
> > +	newPath := ""
> > +	for i := range pathElements {
> > +		newPath = strings.Join([]string{newPath, sep, pathElements[i]}, "")
> > +		if _, err := conn.Client.Create(
> > +			newPath,
> > +			nil,
> > +			0,
> > +			zk.WorldACL(zk.PermAll)); err != nil && err != zk.ErrNodeExists {
> > +			return err
> > +		}
> > +	}
> > +	return nil
> > +}
> > +
> > +func (conn ZooKeeper) registerService(service *discovery.ServiceDefinition) error {
> Is there any shared state being created here under the hood in the Zookeeper struct? Does this need a pointer receiver? (i.e. func (conn *Zookeeper) registerService...)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>